### PR TITLE
Integrating Deployment templates into new generic handler

### PR DESCRIPTION
**Pull request by bfoley13**
# Description
This migrates the Deployment templates into the new Generic Template Handler.  Non-breaking migration that results in same file output as before through CLI or package usage.

Added DeepCopy support for Template/DraftConfig.
Added filename override handling

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Refactor

## How Has This Been Tested?

Provide instructions so we can reproduce, and list any relevant details for your test configuration. Please mention if this is a breaking change which will impact consuming tool(s).

- [x] Local Tests 
- [x] Unit Test

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by Vidya2606 — APPROVED** [review]
lgtm

File: cmd/info.go [hunk 1]
```diff
@@ -18,6 +17,10 @@ const (
 	JSON Format = "json"
 )
 
+var (
```
Comment (Tatsinnit) [inline cmd/info.go:20]: 💡 Thoughts/observation please:

Please note, depending on the usage, you can declare them as a var and treat them as constants in terms of not mutating them:

This makes the length of the collection as fixed at compile time and prevents resizing. If you want to ensure immutability.

```*.go
var supportedDeploymentTypes = [...]string{
	"helm",
	"kustomize",
	"manifests",
}
```

OR, might be your specific equivalent rewrite.

```*.go
var supportedDeploymentTypes = []string{
	"helm",
	"kustomize",
	"manifests",
}
```

File: example/deployment.go [hunk 2]
```diff
@@ -32,76 +18,37 @@ func WriteDeploymentFilesExample() error {
...
+		"GENERATORLABEL": "draft",
 	}
 
 	// Set the output path for the deployment files
 	outputPath := "./"
 
-	// Write the deployment files
-	err := WriteDeploymentFiles(&w, outputPath, deployConfig, deploymentType)
+	// Get the deployment template
+	d, err := handlers.GetTemplate(deploymentTemplateType, "", outputPath, &w)
```
Comment (Tatsinnit) [inline example/deployment.go:38]: 💡 Error_Handling suggestion:

```*.go
// Get the deployment template
d, err := handlers.GetTemplate(deploymentTemplateType, "", outputPath, &w)
if err != nil {
    return fmt.Errorf("failed to get template: %w", err) // Use %w for wrapping errors
}
if d == nil {
    return fmt.Errorf("template is nil")
}

```

File: example/deployment.go [hunk 3]
```diff
@@ -32,76 +18,37 @@ func WriteDeploymentFilesExample() error {
...
+	if err != nil || d == nil {
+		return fmt.Errorf("failed to get template: %e", err)
+	}
+
+	// Set the variable values within the template
+	for k, v := range templateVars {
+		d.Config.SetVariable(k, v)
```
Comment (Tatsinnit) [inline example/deployment.go:45]: 💡 Suggestion/Observation thought share only please:

```*.go
// Set the variable values within the template
for k, v := range templateVars {
    if err := d.Config.SetVariable(k, v); err != nil {
        return fmt.Errorf("failed to set variable %s: %w", k, err)
    }
}

```

File: pkg/handlers/deployment_test.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/Azure/draft/pkg/fixtures"
+	"github.com/Azure/draft/pkg/templatewriter/writers"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestManifestDeploymentValidation(t *testing.T) {
```
Comment (bosesuneha) [inline pkg/handlers/deployment_test.go:14]: Do you think we can move this test to `template_test.go` file?

File: cmd/create.go [hunk 5]
```diff
@@ -304,21 +304,20 @@ func (cc *createCmd) generateDockerfile(langConfig *config.DraftConfig, lowerLan
...
-	d := deployments.CreateDeploymentsFromEmbedFS(template.Deployments, cc.dest)
 	var deployType string
-	var deployConfig *config.DraftConfig
+	var deployTemplate *handlers.Template
 	var err error
 
 	if cc.createConfig.DeployType != "" {
 		deployType = strings.ToLower(cc.createConfig.DeployType)
-		deployConfig, err = d.GetConfig(deployType)
+		deployTemplate, err = handlers.GetTemplate(fmt.Sprintf("deployment-%s", deployType), "", cc.dest, cc.templateWriter)
```
Comment (bosesuneha) [inline cmd/create.go:313]: How are we going to pass the version? Will it be added to the createCmd struct?

File: pkg/config/draftconfig.go [hunk 6]
```diff
@@ -211,6 +225,46 @@ func (d *DraftConfig) VariableMapToDraftConfig(flagVariablesMap map[string]strin
...
+	newConfig.Versions = d.Versions
+	newConfig.DefaultVersion = d.DefaultVersion
+
+	newConfig.Variables = make([]*BuilderVar, len(d.Variables))
+	for i, variable := range d.Variables {
+		newConfig.Variables[i] = variable.DeepCopy()
+	}
+
+	return newConfig
+}
```
Comment (Tatsinnit) [inline pkg/config/draftconfig.go:251]: 💡 Observation/Suggestion only please: struct literal to initialize newConfig fields, making the code more compact and readable. Which means we can jus use the d.variable as is, loop remains same.

```*.go
func (d *DraftConfig) DeepCopy() *DraftConfig {
	newConfig := &DraftConfig{
		TemplateName:   d.TemplateName,
		DisplayName:    d.DisplayName,
		Description:    d.Description,
		Type:           d.Type,
		Versions:       d.Versions,
		DefaultVersion: d.DefaultVersion,
		Variables:      make([]*BuilderVar, len(d.Variables)),
	}

	for i, variable := range d.Variables {
		newConfig.Variables[i] = variable.DeepCopy()
	}

	return newConfig
}

```

File: pkg/config/draftconfig.go [hunk 7]
```diff
@@ -211,6 +225,46 @@ func (d *DraftConfig) VariableMapToDraftConfig(flagVariablesMap map[string]strin
...
+	newVar.Description = bv.Description
+	newVar.Type = bv.Type
+	newVar.Kind = bv.Kind
+	newVar.Value = bv.Value
+	newVar.Versions = bv.Versions
+
+	newVar.ExampleValues = make([]string, len(bv.ExampleValues))
+	copy(newVar.ExampleValues, bv.ExampleValues)
+	return newVar
+}
```
Comment (Tatsinnit) [inline pkg/config/draftconfig.go:266]: 💡 same suggestion for this if you like this style: https://github.com/Azure/draft/pull/393/files#r1792581736

**Review by Tatsinnit — APPROVED** [review]
❤️ Thank you for kind ping, left few suggestions, rest looks awesome!! I am sure other folks will add anything if I've missed. Thanks heaps.

File: cmd/info.go (hunk 1 above)
Comment (bfoley13) [reply cmd/info.go:20]: changed to your first suggestion!

File: example/deployment.go (hunk 2 above)
Comment (bfoley13) [reply example/deployment.go:38]: nice catch, could potentially be a `nil` error logged so more explicit message is better!

File: example/deployment.go (hunk 3 above)
Comment (bfoley13) [reply example/deployment.go:45]: `SetVariable` doesnt have a response on it so no error can be thrown

File: pkg/config/draftconfig.go (hunk 6 above)
Comment (bfoley13) [reply pkg/config/draftconfig.go:251]: cleaned up all the `DeepCopy` definitions

**Review by bosesuneha — APPROVED** [review]

**Pull request merged**